### ISSUE TITLE
Only render updated items in an array

### ISSUE THREAD

**Issue by ryanking1809** [react-spring/zustand#63]
I have an array of items in Zustand store, and if I update one, all items in the array rerender.

I have a simple example here where a store has a list of colors rendered to divs. When you click on a div it updates the items in the stored array. However, all items rerender when one is changed. Is there a way to set up Zustand so only one object rerenders when it item changes in an array?

https://codesandbox.io/s/polished-framework-59eys
```jsx
import React from "react";
import ReactDOM from "react-dom";
import produce from "immer";
import create from "zustand";

export const immer = config => set => {
  return Object.entries(config()).reduce(
    (acc, [key, value]) => ({
      ...acc,
      [key]:
        typeof value === "function"
          ? (...args) => set(produce(draft => void config(draft)[key](...args)))
          : value
    }),
    {}
  );
};

const store = state => ({
  divs: ["red", "red", "red"],
  updateColor: (color, index) => {
    state.divs[index] = color;
  }
});

const [useStore] = create(immer(store));

const App = () => {
  const { divs } = useStore();
  return (
    <div className="App">
      {divs.map((d, i) => (
        <DivEl color={d} index={i} />
      ))}
    </div>
  );
};

const DivEl = ({ color, index }) => {
  const { updateColor } = useStore();
  return (
    <div
      onClick={() => updateColor(color === "red" ? "blue" : "red", index)}
      style={{
        width: "100px",
        height: "100px",
        background: color,
        margin: "1em"
      }}
    />
  );
};

const rootElement = document.getElementById("root");
ReactDOM.render(<App />, rootElement);
```

**Comment by Magellol** [comment]
I think it might be because you don't have a key on each element. React uses keys to know which item has changed or has been removed/added to the list. Without a stable identity, it doesn't know and renders everything again.

```diff
<div className="App">
  {divs.map((d, i) => (
-    <DivEl color={d} index={i} />
+    <DivEl color={d} index={i} key={i} />
  ))}
</div>
```

Now using indexes as keys is not always good, it depends if the items in the list will have their order changed eventually. In that case you might need something that is uniquely tied up to each item.

**Comment by ryanking1809** [comment]
Sorry, that was clumsy of me, but the keys don't make a difference. I've updated the codesandbox above and all items still rerender when only one item has changed.

**Comment by Magellol** [comment]
Sorry I got confused as well :). Not having stable keys would actually force React to destroy and create items on each render pass because it doesn't know which items got updated/added/removed since the keys are different (or no keys at all).

The fact that all `DivEl` re-renders is normal because it's parent re-renders due to the `Divs` state slice to update in the store. To avoid unnecessary renders, we can leverage `memo` which would do a shallow equal comparison on the props and abort rendering if they haven't changed.

I forked your sandbox and used memo: https://codesandbox.io/s/musing-curran-gwnor
